### ISSUE TITLE
problem with controller,

### ISSUE THREAD

**Issue by FrostKnight** [fgsfdsfgs/perfect_dark#518]
Seems you don't have the raphnet 3.6 version of the n64 controller(s) dual and single.

I keep trying to play, but some configurations arent working such as Z.

But yeah, mupen64plus 2.5.9 has some control options you are missing one of them is the version above.

**Comment by fgsfdsfgs** [comment]
What do mupen64 or the raphnet plugin have to do with it? I don't quite understand.
You can try downloading [this file](https://raw.githubusercontent.com/mdqinc/SDL_GameControllerDB/refs/heads/master/gamecontrollerdb.txt) and putting it in the game folder (or whatever pd.ini is on your machine).

**Comment by FrostKnight** [comment]
> What do mupen64 or the raphnet plugin have to do with it? I don't quite understand. You can try downloading [this file](https://raw.githubusercontent.com/mdqinc/SDL_GameControllerDB/refs/heads/master/gamecontrollerdb.txt) and putting it in the game folder (or whatever pd.ini is on your machine).

Will try later then

**Comment by FrostKnight** [comment]
> What do mupen64 or the raphnet plugin have to do with it? I don't quite understand. You can try downloading [this file](https://raw.githubusercontent.com/mdqinc/SDL_GameControllerDB/refs/heads/master/gamecontrollerdb.txt) and putting it in the game folder (or whatever pd.ini is on your machine).

Tried it, didn't work for some reason. If you need to know what it needs to look like, I will tell you:

https://github.com/mupen64plus/mupen64plus-input-sdl/blob/master/data/InputAutoCfg.ini

scroll down to this:

; raphnet technologies USB converter, v3 series (http://www.raphnet-tech.com/

**Comment by FrostKnight** [comment]
> What do mupen64 or the raphnet plugin have to do with it? I don't quite understand. You can try downloading [this file](https://raw.githubusercontent.com/mdqinc/SDL_GameControllerDB/refs/heads/master/gamecontrollerdb.txt) and putting it in the game folder (or whatever pd.ini is on your machine).

Raphnet plugin has newer control options in 2.5.9 tnen 2.5.

That's why it matters. just to clarify.  V3 of raphnet's plugin doesn't exist in yours currently. Just the v1-v2.

And they function much differently.

**Comment by FrostKnight** [comment]
> What do mupen64 or the raphnet plugin have to do with it? I don't quite understand. You can try downloading [this file](https://raw.githubusercontent.com/mdqinc/SDL_GameControllerDB/refs/heads/master/gamecontrollerdb.txt) and putting it in the game folder (or whatever pd.ini is on your machine).

So you said I should download that and rename it pd.ini or something else? Also, I suppose I could ask them to add the config if it isnt there already

**Comment by fgsfdsfgs** [comment]
Don't rename it. Put it in the directory where `pd.ini` is.

**Comment by FrostKnight** [comment]
> Don't rename it. Put it in the directory where `pd.ini` is.

There seems to be some sort of bug, Z button, start button and control stick all don't work with this using v3 raphnet n64.

Although, I noticed in that config you told me to download that windows has that config but not linux.

**Comment by FrostKnight** [comment]
Yeah... I tried to get help, from the person, but he gave me confusing instructions. You  can see issue 723  on his issue tracker for more info.

However, it turns out when I tried to start the game, previously last week V3 of raphnet's plugin was in fact selected, but it doesn't quite work right. L, R and Z do nothing.

Launched pd like this:

./pd 

in the correct folder and found this out from the terminal

**Comment by FrostKnight** [comment]
This problem is driving me kind of nuts, could you at some point make it an option to make it so I can use this:

https://github.com/mupen64plus/mupen64plus-input-sdl/blob/master/data/InputAutoCfg.ini

instead of this:

https://raw.githubusercontent.com/mdqinc/SDL_GameControllerDB/refs/heads/master/gamecontrollerdb.txt

The latter is useless for L, R and Z buttons.

Also, the way to set it up is massively confusing for classic n64 controllers. The ones that Nintendo originally released with n64 all those years ago.

Basically, I need n64 style controller bindings.

**Comment by fgsfdsfgs** [comment]
Well, you seem to be the only person with this problem so far and I'm not really sure how a different config format will help.
You can also probably convert that config to the normal SDL binding format. Try pasting something like this to the end of your `gamecontrollerdb.txt`:
```
030000009b2800003200000000000000,Raphnet GC and N64 Adapter,a:b0,b:b7,dpdown:b11,dpleft:b12,dpright:b13,dpup:b10,lefttrigger:+a5,leftx:a0,lefty:a1,rightshoulder:b2,righttrigger:+a2,rightx:a3,righty:a4,start:b3,x:b1,y:b8,platform:Linux,
030000009b2800006000000000000000,Raphnet GC and N64 Adapter,a:b0,b:b7,dpdown:b11,dpleft:b12,dpright:b13,dpup:b10,lefttrigger:+a5,leftx:a0,lefty:a1,rightshoulder:b2,righttrigger:+a2,rightx:a3,righty:a4,start:b3,x:b1,y:b8,platform:Linux,
030000009b2800006100000000000000,Raphnet N64 Adapter,+rightx:b9,+righty:b7,-rightx:b8,-righty:b6,a:b0,b:b1,dpdown:b11,dpleft:b12,dpright:b13,dpup:b10,leftshoulder:b4,lefttrigger:b2,leftx:a0,lefty:a1,rightshoulder:b5,start:b3,platform:Linux,
030000009b2800006300000000000000,Raphnet N64 Adapter,+rightx:b9,+righty:b7,-rightx:b8,-righty:b6,a:b0,b:b1,dpdown:b11,dpleft:b12,dpright:b13,dpup:b10,leftshoulder:b4,lefttrigger:b2,leftx:a0,lefty:a1,rightshoulder:b5,start:b3,platform:Linux,
```
Z Trigger should then map to `JOYn_RTRIGGER` in the menus.

**Comment by FrostKnight** [comment]
> Well, you seem to be the only person with this problem so far and I'm not really sure how a different config format will help. You can also probably convert that config to the normal SDL binding format. Try pasting something like this to the end of your `gamecontrollerdb.txt`:
> 
> ```
> 030000009b2800003200000000000000,Raphnet GC and N64 Adapter,a:b0,b:b7,dpdown:b11,dpleft:b12,dpright:b13,dpup:b10,lefttrigger:+a5,leftx:a0,lefty:a1,rightshoulder:b2,righttrigger:+a2,rightx:a3,righty:a4,start:b3,x:b1,y:b8,platform:Linux,
> 030000009b2800006000000000000000,Raphnet GC and N64 Adapter,a:b0,b:b7,dpdown:b11,dpleft:b12,dpright:b13,dpup:b10,lefttrigger:+a5,leftx:a0,lefty:a1,rightshoulder:b2,righttrigger:+a2,rightx:a3,righty:a4,start:b3,x:b1,y:b8,platform:Linux,
> 030000009b2800006100000000000000,Raphnet N64 Adapter,+rightx:b9,+righty:b7,-rightx:b8,-righty:b6,a:b0,b:b1,dpdown:b11,dpleft:b12,dpright:b13,dpup:b10,leftshoulder:b4,lefttrigger:b2,leftx:a0,lefty:a1,rightshoulder:b5,start:b3,platform:Linux,
> 030000009b2800006300000000000000,Raphnet N64 Adapter,+rightx:b9,+righty:b7,-rightx:b8,-righty:b6,a:b0,b:b1,dpdown:b11,dpleft:b12,dpright:b13,dpup:b10,leftshoulder:b4,lefttrigger:b2,leftx:a0,lefty:a1,rightshoulder:b5,start:b3,platform:Linux,
> ```
> 
> Z Trigger should then map to `JOYn_RTRIGGER` in the menus.

I just had an idea... how would I change the PD.ini to make this work. Your idea didn't, but I tried tweaking pd.ini and I got L to work. Can you give me some advice?

Like what buttons I would need to change for input player 1... etc

**Comment by FrostKnight** [comment]
Do you know of anyone who can use this with the regular n64 controller?

Mine is not wireless, hence my curiosity...

I don't particularly like beating a dead horse, but I actually wonder if anyone has gotten raphnet v3, or for that matter any n64 config to work that is using the standard n64 button setup.

If you or anyone can tell me, please do so.

**Comment by FrostKnight** [comment]
Been meaning to mention, I am so serious about wanting mupen64plus config being an option to make controllers work, the InputAutoCfg.ini one, that I would probably donate money if that's what it took to convince you. I am aware you might someday add online option, thus, this does interest me a lot.

I think most people haven't complained because they are using wireless or non-standard n64 controllers. Anyways let me know if donations would interest you to get this going.

**Comment by LonelySpaceDetective** [comment]
> I think most people haven't complained because they are using wireless or non-standard n64 controllers.

This is a tad off-topic, but if you want my input, I think what's happening here is more you have a Venn diagram of overlapping problems here.

One is you're dealing with niche software that is mostly developed by a single person, another is you're using what sounds like an unusual controller setup like you mentioned here (also honestly, I imagine "most" players are probably not using N64 controllers at all with this port; either KB+M or a standard modern gamepad), you're on Linux which is not as popular an OS as Windows and more fiddly in general, annnnnnd... I get the impression you're not very tech savvy in general? I'm not trying to demean you or anything, but the impression I get from some of your posts and bug reports that you don't work very well with computers and have issues explaining technical issues or following instructions when they involve technical tasks.
In fact, the last two (Linux and not being tech savvy) being combined is probably the big issue here.

A lot of people have at least one of these issues, but a single issue can often be overcome with enough help or persistence/knowledge, while you having all of these at once makes it a harder task for someone to try to help you through this; for instance I would normally try to guide you through this, but between you being on Linux and you using real N64 controllers with an adapter I don't know how I would even start troubleshooting this.

I'm not trying to hate on you or anything by saying this, just that I'm saying you're in a bad position to be having technical problems between all those factors. You might honestly have to wait until someone _else_ on Linux and who uses N64 controllers with the Raphnet adapter has the same problem as you, or just settle for using a normal Xbox gamepad or something (or, I guess, switch to Windows where you might be able to get more help).

**Comment by FrostKnight** [comment]
> > I think most people haven't complained because they are using wireless or non-standard n64 controllers.
> 
> This is a tad off-topic, but if you want my input, I think what's happening here is more you have a Venn diagram of overlapping problems here.
> 
> One is you're dealing with niche software that is mostly developed by a single person, another is you're using what sounds like an unusual controller setup like you mentioned here (also honestly, I imagine "most" players are probably not using N64 controllers at all with this port; either KB+M or a standard modern gamepad), you're on Linux which is not as popular an OS as Windows and more fiddly in general, annnnnnd... I get the impression you're not very tech savvy in general? I'm not trying to demean you or anything, but the impression I get from some of your posts and bug reports that you don't work very well with computers and have issues explaining technical issues or following instructions when they involve technical tasks. In fact, the last two (Linux and not being tech savvy) being combined is probably the big issue here.
> 

Perhaps your right, but I do wonder what person would play this game without an n64 controller.

Its so much harder to do it with a gamepad like interface because it has two control sticks usually.

Also, I am tech savy on some things, but not something this confusing. Its just too confusing to try the sdl2 input that they suggested. And the Z button, L and R both don't work for this for some reason and yes I use linux.

> A lot of people have at least one of these issues, but a single issue can often be overcome with enough help or persistence/knowledge, while you having all of these at once makes it a harder task for someone to try to help you through this; for instance I would normally try to guide you through this, but between you being on Linux and you using real N64 controllers with an adapter I don't know how I would even start troubleshooting this.

True....
> 
> I'm not trying to hate on you or anything by saying this, just that I'm saying you're in a bad position to be having technical problems between all those factors. You might honestly have to wait until someone _else_ on Linux and who uses N64 controllers with the Raphnet adapter has the same problem as you, or just settle for using a normal Xbox gamepad or something (or, I guess, switch to Windows where you might be able to get more help).


That is what I was afraid of. It would be nice if someone could advertise a fix for this. I tried to ask on reddit, but people were obnoxious and acted like they had the right to not answer  just to be aggravating.  This sometimes happens on linux forums and/or similar.  People act snotty... not that this here is what I am hearing from you guys on this thread. ;)

**Comment by fgsfdsfgs** [comment]
Unfortunately at this point I don't really know how to fix it, since you apparently tried everything, like using different SDL controller configs and generating bindings yourself using GamepadTool or whatever. Since you're using Linux I don't think you have any option of remapping it outside of the game either, or at least none known to me. All I do know is that other people with Raphnet adapters had them working fine and at least one of them was also running Linux, so they should be working fine by that logic.

Using mupen64's config files is not really going to help here, since they essentially contain the same thing that SDL_gamecontrollerdb does, just in a different format.

However, I've committed something that might help narrow down the issue. The game will now print the controller GUID in the terminal when it detects the controller. Could you run the newest build (after it's done and uploaded in like 15 minutes) and tell me what GUID the game prints for your Raphnet adapter?

**Comment by LonelySpaceDetective** [comment]
> Perhaps your right, but I do wonder what person would play this game without an n64 controller.
> 
> Its so much harder to do it with a gamepad like interface because it has two control sticks usually.

I mean, a lot of people find that a second stick makes FPSes _easier_ to play, myself included. Though you do echo some people I've seen on the Steam Discussions forum for the recently released Nightdive remaster of the classic DOOM games, who want an option for something more like the original PS1 port's layout. Not that I really understand them, but you're not alone.

And I admit, I can't say for certain what controllers are most commonly used without analytics (which even then they would probably be misleading due to people using DirectInput>XInput wrappers), but keyboard and mouse requires no setup whatsoever and is by a big margin the preferred control method for FPSes on PC, and XInput-compatible controllers generally require only the setup of plugging them in or hitting the start button, with Xbox controllers in particular considered a gold standard for gamepads on PC.
As niche software that ports an N64 game to modern platforms, there probably _is_ more people using N64 controllers than there would normally be I'll give you, but I imagine if you were to poll the playerbase, I'd wager people using N64 controllers would probably still be a distant third place at highest; and behind XInput controllers and KB+M.

Like, I wouldn't think the majority of people playing e.g. the PC ports of Super Mario 64 or Ocarina of Time are using their old N64 controllers from the '90s.

Anyway, I'm gonna stop with the tangent spamming this issue thread now. I hope you're able to find or figure out a solution for your issue.

**Comment by FrostKnight** [comment]
> > Perhaps your right, but I do wonder what person would play this game without an n64 controller.
> > Its so much harder to do it with a gamepad like interface because it has two control sticks usually.
> 
> I mean, a lot of people find that a second stick makes FPSes _easier_ to play, myself included. Though you do echo some people I've seen on the Steam Discussions forum for the recently released Nightdive remaster of the classic DOOM games, who want an option for something more like the original PS1 port's layout. Not that I really understand them, but you're not alone.
> 
> And I admit, I can't say for certain what controllers are most commonly used without analytics (which even then they would probably be misleading due to people using DirectInput>XInput wrappers), but keyboard and mouse requires no setup whatsoever and is by a big margin the preferred control method for FPSes on PC, and XInput-compatible controllers generally require only the setup of plugging them in or hitting the start button, with Xbox controllers in particular considered a gold standard for gamepads on PC. As niche software that ports an N64 game to modern platforms, there probably _is_ more people using N64 controllers than there would normally be I'll give you, but I imagine if you were to poll the playerbase, I'd wager people using N64 controllers would probably still be a distant third place at highest; and behind XInput controllers and KB+M.
> 
> Like, I wouldn't think the majority of people playing e.g. the PC ports of Super Mario 64 or Ocarina of Time are using their old N64 controllers from the '90s.
> 
> Anyway, I'm gonna stop with the tangent spamming this issue thread now. I hope you're able to find or figure out a solution for your issue.

Last thing I will say, having two sticks is terrible for perfect dark or other FPS games. Its way harder to strafe accurately.

Strafing makes you move twice as fast, but if you had a 2nd stick instead of the C buttons or w/e, you make it even more difficult to manuever using it.

That's why I do it this way. 

anywho, yeah no easy answer here.

I hope someone fixes this issue. I cannot.

**Comment by Specker101** [comment]
Hi FrostKnight,
I'm using the Raphnet N64/GC dual adapter with an original controller and rumble pack (built the adapter myself with an Arduino).
I'm not exactly sure where the problem is on your end, but I just want to tell you what I did.

First, I wrote this in the gamecontrollerdb.txt file:

030000009b2800006300000000000000,Dual GC/N64 to USB v3.6,platform:Windows,a:b0,b:b1,start:b3,dpup:b6,dpdown:b7,dpleft:b8,dpright:b9,rightx:a0,righty:a1,lefttrigger:b5,righttrigger:b2,

Be sure to delete everything else related to Raphnet N64 in this file.

In the pd.ini, I just changed:
CANCEL_BUTTON=JOY1_B

And in-game, you have to change the control style to 1.1. 
I forgot this at first and had similar problems like you.

With this, I was able to use the controller as intended. Only rumble won’t work this way.

For rumble support, I'm now using the tool XOutput.
With this tool, I have rumble with the original rumble pack and don’t need the gamecontrollerdb stuff anymore.
But you need to crank up the rumble setting in-game — otherwise the rumble pack won’t do anything.

Maybe you’ll have to search for a XOutput Linux equivalent, or install a Windows partition just for gaming?

**Comment by FrostKnight** [comment]
Problem is, I use the classic nintendo 64 controller, this has C up, C down, C left, C right instead of an additional control stick

I hope he will at least change this to being a low priority or medium priority in fixing.

Would need that mupen64plus 2.5.9 inputs for it to work though.

Would be willing to pay a small fee for it being added.

**Comment by Specker101** [comment]
Did you read that I’m using the classic controller as well?
Did you try my solution?
It’s working just fine for me.

**Comment by FrostKnight** [comment]
> Did you read that I’m using the classic controller as well? Did you try my solution? It’s working just fine for me.

Are you using linux, raphnet v3 controller adapter and a classic controller? I guess ill try again?

What did you change this to:

CANCEL_BUTTON=JOY1_B

???

**Comment by Specker101** [comment]
I installed Linux today, just to test it for you (and because I find these problems interesting).
My takeaway after two hours of tinkering: more or less the same experience you had.
The Z button is not recognized (the start button works, though).
I tested the SDL_GamePad tool, but it seems to have no effect under Linux.
I'm not giving up — if I find a solution, I'll let you know!

**Comment by FrostKnight** [comment]
> I installed Linux today, just to test it for you (and because I find these problems interesting). My takeaway after two hours of tinkering: more or less the same experience you had. The Z button is not recognized (the start button works, though). I tested the SDL_GamePad tool, but it seems to have no effect under Linux. I'm not giving up — if I find a solution, I'll let you know!
The Z button isn't the only problem though.

A few other keys are not behaving right either somehow.

I need all of the keys to work correctly. I cannot remember specifics right now, but yeah its more than just Z.  To get to finding my problem you might have to use devuan daedalus or debian bookworm. That's what I was using when I had those problems.

Although its possible other linux distros have this problem too.

I was using the devuan option though.

**Comment by Specker101** [comment]
I got it to work now. That means you can also get it to work. I'll show you how I did it.

"A few other keys are not behaving right either somehow."
Yes, same behavior on my system.

The problem was that the SDL2 Gamepad Tool mappings had no effect whatsoever at first.
Neither a gamecontrollerdb.txt file nor setting the mapping as an environment variable worked. (probably only on windows)
However, with the Gamepad Tool, I already saw that the app recognized all buttons on the original N64 controller via the Raphnet adapter.

So what we must do instead is the following:

1. Start the game from the terminal with ./pd.x86_64.

2. Look out for "Input: GUID for controller".

3. Copy the GUID string — in my case, it’s 0300497f9b2800006300000001010000.

4. Start the game from the terminal with this command (change your GUID if you have a different one):
   SDL_GAMECONTROLLERCONFIG="0300497f9b2800006300000001010000,raphnet technologies Dual GC/N64 to USB v3.6,platform:Linux,a:b0,b:b1,start:b3,leftshoulder:b4,dpup:b6,dpdown:b7,dpleft:b8,dpright:b9,rightx:a0,righty:a1,lefttrigger:b5,righttrigger:b2,crc:7f49," ./pd.x86_64

Now you are able to map all buttons in-game!

**Comment by FrostKnight** [comment]
> I got it to work now. That means you can also get it to work. I'll show you how I did it.
> 
> "A few other keys are not behaving right either somehow." Yes, same behavior on my system.
> 
> The problem was that the SDL2 Gamepad Tool mappings had no effect whatsoever at first. Neither a gamecontrollerdb.txt file nor setting the mapping as an environment variable worked. (probably only on windows) However, with the Gamepad Tool, I already saw that the app recognized all buttons on the original N64 controller via the Raphnet adapter.
> 
> So what we must do instead is the following:
> 
>     1. Start the game from the terminal with ./pd.x86_64.
> 
>     2. Look out for "Input: GUID for controller".
> 
>     3. Copy the GUID string — in my case, it’s 0300497f9b2800006300000001010000.
> 
>     4. Start the game from the terminal with this command (change your GUID if you have a different one):
>        SDL_GAMECONTROLLERCONFIG="0300497f9b2800006300000001010000,raphnet technologies Dual GC/N64 to USB v3.6,platform:Linux,a:b0,b:b1,start:b3,leftshoulder:b4,dpup:b6,dpdown:b7,dpleft:b8,dpright:b9,rightx:a0,righty:a1,lefttrigger:b5,righttrigger:b2,crc:7f49," ./pd.x86_64
> 
> 
> Now you are able to map all buttons in-game!

Two things, what you said worked! 

But for some odd reason I didn't see my GUID string lol.

But irrelevant since it works anyways, thank you. Could you perhaps add this information to here:

https://github.com/mdqinc/SDL_GameControllerDB

So that other people will not have this damned issue?
 
:)

**Comment by fgsfdsfgs** [comment]
The gamecontrollerdb file should work on all platforms, though on Linux you probably have to place it in the save folder. However, the format of lines in that file has changed since SDL2 Gamepad Tool was last updated, so maybe that's why we couldn't figure it out earlier. Well, in any case, glad you guys figured it out, because I had no ideas remaining.

**Comment by FrostKnight** [comment]
not sure what is happening, but now its not working for me anymore. Its very peculiar. The Z button works using:

 SDL_GAMECONTROLLERCONFIG="0300497f9b2800006300000001010000,raphnet technologies Dual GC/N64 to USB v3.6,platform:Linux,a:b0,b:b1,start:b3,leftshoulder:b4,dpup:b6,dpdown:b7,dpleft:b8,dpright:b9,rightx:a0,righty:a1,lefttrigger:b5,righttrigger:b2,crc:7f49," ./pd.x86_64

but now the + pad doesn't work correctly. I don't know what I did... wrong.
